File: hto/14/hto_14.cpp
#include <iostream>
#include <string>
#include <vector>
#include <climits>
#include <algorithm>
#include <cstring>
using namespace std;

class Solution {
public:
    int cuttingRope(int n) {
        vector<int> dp(n+1, 0);
        dp[0] = 1;
        for (int i = 1; i <= (n+1)/2; ++i) {
            for (int j = i; j <= n; ++j) {
                dp[j] = max(dp[j], i * dp[j-i]);
            }
        }
        return dp[n];
    }
};

int main() {
    Solution s;
    cout << s.cuttingRope(10) << endl;
}

File: hto/14/hto_14.go
func cuttingRope(n int) int {
    dp := make([]int, n+1)
    dp[2] = 1
    for i := 3; i <= n; i++ {
        for j := 2; j < i; j++ {
            dp[i] = max(dp[i], max(j * (i-j), j * dp[i-j]))
        }
    }
    return dp[n]
}

func max(a int, b int) int {
    if a > b {
        return a
    }
    return b
}
